app/models/user.rb

```diff
@@ -40,20 +46,27 @@ class User < ActiveRecord::Base
     CONTESTER
   ]
 
-  # Authenticates a user by their login name and unencrypted password.  Returns the user or nil.
-  #
-  # uff.  this is really an authorization, not authentication routine.
-  # We really need a Dispatch Chain here or something.
-  # This will also let us return a human error message.
-  #
-  def self.authenticate(login, password)
-    user = find_by_login(login.downcase) || find_by_email(login.downcase) # need to get the salt
-
-    if user and user.password == encrypt_password(password)
-      return user
+  def self.from_omniauth(auth)
+    pass = Devise.friendly_token[0,20]
+    where(provider: auth.provider, uid: auth.uid).first_or_create do |user|
+      user.provider = auth.provider
+      user.uid = auth.uid
+      user.email = auth.info.email
+      user.name = auth.info.name
+      user.login = auth.info.first_name
```

Comment:
Let's use the `email` as `login` here. This has the meaning of a username for logging in.